protocol-units/bridge/contracts/src/AtomicBridgeInitator.sol

```diff
@@ -0,0 +1,113 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.22;
+
+import {IAtomicBridgeInitiator} from "./IAtomicBridgeInitiator.sol";
+import {IWETH10} from "./WETH/interfaces/IWETH10.sol";
+import {Initializable} from "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+
+contract AtomicBridgeInitiator is IAtomicBridgeInitiator, Initializable {
+    struct BridgeTransfer {
+        uint256 amount;
+        address originator;
+        bytes32 recipient;
+        bytes32 hashLock;
+        uint256 timeLock;
+        bool completed;
+    }
+
+    mapping(bytes32 => BridgeTransfer) public bridgeTransfers;
+    IWETH10 public weth;
+    address public authority;
+
+    modifier onlyAuthorized() {
+        if (msg.sender != authority) revert Unauthorized();
+        _;
+    }
+
+    function initialize(address _weth) public initializer {
+        if (_weth == address(0)) {
+            revert ZeroAddress();
+        }
+        authority = msg.sender;
+        weth = IWETH10(_weth);
+    }
+
+    function initiateBridgeTransfer(uint256 _wethAmount, bytes32 _recipient, bytes32 _hashLock, uint256 _timeLock)
+        external
+        payable
+        returns (bytes32 _bridgeTransferId)
+    {
+        address originator = msg.sender;
+        uint256 ethAmount = msg.value;
+        uint256 totalAmount = _wethAmount + ethAmount;
+        _bridgeTransferId = keccak256(abi.encodePacked(originator, _recipient, _hashLock, _timeLock, block.timestamp));
+        // Check if the bridge transfer already exists
+        if (bridgeTransfers[_bridgeTransferId].amount != 0) revert BridgeTransferExists();
+        // Ensure there is a valid total amount
+        if (totalAmount == 0) {
+            revert ZeroAmount();
+        }
+        // If msg.value is greater than 0, convert ETH to WETH
+        if (ethAmount > 0) weth.deposit{value: ethAmount}();
+
+        //Transfer WETH to this contract, revert if transfer fails
+        if (_wethAmount > 0) {
+            if (!weth.transferFrom(originator, address(this), _wethAmount)) revert WETHTransferFailed();
+        }
+        bridgeTransfers[_bridgeTransferId] = BridgeTransfer({
+            amount: totalAmount,
+            originator: originator,
+            recipient: _recipient,
+            hashLock: _hashLock,
+            timeLock: block.timestamp + _timeLock,
+            completed: false
```

Comment:
Can we also check if the bridge implementation allows us to retrieve a list of all bridgeTransferIds from the bridgeTransfers map in the contract? I need this to rebuild the state of the bridge if it goes down. If we cannot iterate over the stored bridgeTransfers, then keeping past transfers may not be necessary. This is an important question to answer as it is crucial for rebuilding the state of the Bridge Service.

Reply:
I looked into this. There is no way solidiy allows you to get all the keys from the mapping, so the only way is to store a seperate array of all the keys, then write a function
```
function getAllBridgeTransferIds() external view returns (bytes32[] memory) 
```
We need the getter function because the inbuilt get method on an array in solidity can only get a single element in the array by index. The getter function I've added will return the entire array.